docs/proposals/affinity.md

```diff
@@ -0,0 +1,341 @@
+of this document.
+
+## API
+
+We will add `AffinitySelector` to `PodSpec`.
+
+```go
+
+AffinitySelector *AffinitySelector `json:"affinitySelector,omitempty"`
+
+type AffinitySelector struct {
+	// When there are multiple elements, the lists of machines corresponding to each
+	// AffinityTerm are intersected, i.e. all terms must be satisfied.
+	HardAffinityExpressions []AffinityTerm `json:"hardAffinityExpression,omitempty"`
+	// When there are multiple elements, the lists of machines corresponding to each
+	// AffinityTerm are intersected, i.e. all terms must be satisfied.
+	HardAntiAffinityExpressions []AffinityTerm `json:"hardAntiAffinityExpression,omitempty"`
+	// Satisfying more expressions is better than satisfying fewer expressions.
+	SoftAffinityExpressions []AffinityTerm `json:"softAffinityExpression,omitempty"`
+	// Satisfying more expressions is better than satisfying fewer expressions.
+	SoftAntiAffinityExpressions []AffinityTerm `json:"softAntiAffinityExpression,omitempty"`
+}
+
+type AffinityTerm struct {
+	PodSelector *PodSelector `json:"podSelector,omitempty"` // empty PodSelector matches all pods
+	TopologyKey string `json:"topologyKey,omitempty"`
+}
+```
+
+One caveat: As noted above, the normal semantics of empty `PodSelector` is to match all pods.
+However, `PodSelector` is scoped to namespace, therefore the normal semantics do not allow
+us to match all pods globally. We need to special-case interpret empty `PodSelector` in this
+context to mean "all pods globally."
+
+To explain how this API works, let's say that the `PodSpec` of a pod `P` has an `AffinitySelector`
+that is configured as follows:
+
+```go
+AffinitySelector {
+	HardAffinityExpressions: {{PodSelector: P1, TopologyKey: "node"}},
+	HardAntiAffinityExpressions: {{PodSelector: P2, TopologyKey: "rack"}},
+	SoftAffinityExpressions: {{PodSelector: P3, TopologyKey: "zone"}},
+	SoftAntiAffinityExpressions: {{PodSelector: P4, TopologyKey: "power"}}
+}
+```
+
+Then when scheduling pod P, the scheduler
+* Can only schedule P onto nodes that are running pods that satisfy `P1`. (Assumes all nodes have a label with key "node" and value specifying their node name.)
+* Cannot schedule P onto any racks that are running pods that satisfy `P2`. (Assumes all nodes have a label with key "rack" and value specifying their rack name.)
+* Should try to schedule P onto zones that are running pods that satisfy `P3`. (Assumes all nodes have a label with key "zone" and value specifying their zone.)
+* Should try not to schedule P onto any power domains that are running pods that satisfy `P4`. (Assumes all nodes have a label with key "power" and value specifying their power domain.)
+
+When `HardAffinityExpressions` has multiple elements, the requirements are ANDed. Likewise for `HardAntiAffinityExpressions`.
+For `SoftAffinityExpressions` and `SoftAntiAffinityExpressions`, nodes that satisfy more of the rules are preferred over those that satisfy fewer.
+
+## A comment on symmetry
```

Comment:
In order to craft expressions that are sane we may want to call out the creation of an analysis tool that can run as if it were the scheduler.  

e.g. you have 0 possible matches.  
xref: https://research.cs.wisc.edu/htcondor/HTCondorWeek2013/presentations/KnoellerJ_QAnalyze.pdf